### PR TITLE
Charge native inline within budget

### PR DESCRIPTION
**Pull request by oxade**
## Motivation

This PR sets the budget for native function and breaks down charging to ensure the native functions don't do work which the client cannot afford.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes
## Test Plan

N/A

### CODE REVIEW

File: language/move-vm/runtime/src/native_functions.rs [hunk 1]
```diff
@@ -172,4 +176,35 @@ impl<'a, 'b> NativeContext<'a, 'b> {
...
+    }};
+}
+
+/// Total cost of a native call so far
+#[macro_export]
+macro_rules! native_gas_total_cost {
+    ($native_context:ident, $gas_left:ident) => {{
+        // Its okay to unwrap because the budget can never be less than the gas left
+        $native_context.gas_budget().checked_sub($gas_left).unwrap()
+    }};
```
Comment (dariorussi) [inline language/move-vm/runtime/src/native_functions.rs:209]: so the model is that 
1. we grab all gas left on entry (say gas budget `100`) 
2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.

Is that right?

**Review by dariorussi — APPROVED** [review]
Looks good, I left a comment to make sure I understood correctly

File: language/move-vm/runtime/src/native_functions.rs (hunk 1 above)
Comment (oxade) [reply language/move-vm/runtime/src/native_functions.rs:209]: > so the model is that
> 
> 1. we grab all gas left on entry (say gas budget `100`)
> 2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
> 3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.
> 
> Is that right?

Precisely!

Comment (oxade) [reply language/move-vm/runtime/src/native_functions.rs:209]: > so the model is that
> 
> 1. we grab all gas left on entry (say gas budget `100`)
> 2. we call `native_charge_gas_early_exit` for operations we want to meter, and if that goes out of gas we stop. So say we call it twice with 10, we end up with `gas_left` being 80. If we pass `100` we fail right there
> 3. at the end of a native we call `native_gas_total_cost` to report the cost outside the function into the interpreter. And the cost is budget - whatever is left, in the example 100 - 80 = 20 which is the cost.
> 
> Is that right?

Now on further thought, I can the `gas_left` inside `NativeContext`, which will make things cleaner.
I can tackle this in a subsequent PR

**Pull request merged**